### ISSUE TITLE
HTTP library sets `Content-Length: 0` in every request

### ISSUE THREAD

**Issue by edsko** [well-typed/hackage-security#103]
Relevant section from the spec:

> A message-body MUST NOT be included in a request if the specification of the request method (section 5.1.1) does not allow sending an entity-body in requests.

and

> The presence of a message-body in a request is signaled by the inclusion of a `Content-Length` or `Transfer-Encoding` header field in the request's message-headers.

**Comment by hvr** [comment]
The culprit is in http://hackage.haskell.org/package/HTTP-4000.2.20/docs/src/Network-HTTP-Base.html#mkRequest

```haskell
-- | 'mkRequest method uri' constructs a well formed
-- request for the given HTTP method and URI. It does not
-- normalize the URI for the request _nor_ add the required 
-- Host: header. That is done either explicitly by the user
-- or when requests are normalized prior to transmission.
mkRequest :: BufferType ty => RequestMethod -> URI -> Request ty
mkRequest meth uri = req
 where
  req = 
    Request { rqURI      = uri
            , rqBody     = empty
            , rqHeaders  = [ Header HdrContentLength "0"
                           , Header HdrUserAgent     defaultUserAgent
                           ]
            , rqMethod   = meth
            }

  empty = buf_empty (toBufOps req)
```

**Comment by hvr** [comment]
And here's a trivial fix:

```diff
--- a/hackage-security-HTTP/src/Hackage/Security/Client/Repository/HttpLib/HTTP.hs
+++ b/hackage-security-HTTP/src/Hackage/Security/Client/Repository/HttpLib/HTTP.hs
@@ -69,6 +69,7 @@ get :: Throws SomeRemoteError
 get browser reqHeaders uri callback = wrapCustomEx $ do
     response <- request browser
       $ setRequestHeaders reqHeaders
+      $ removeHeader HTTP.HdrContentLength
       $ HTTP.mkRequest HTTP.GET uri
     case HTTP.rspCode response of
       (2, 0, 0) -> withResponse response callback
@@ -294,3 +295,8 @@ getResponseHeaders response = concat [
     | HTTP.findHeader HTTP.HdrContentEncoding response == Just "gzip"
     ]
   ]
+
+removeHeader :: HTTP.HasHeaders a => HTTP.HeaderName -> a -> a
+removeHeader name h = HTTP.setHeaders h newHeaders
+  where
+    newHeaders = [ x | x@(HTTP.Header n _) <- HTTP.getHeaders h, name /= n ]
```

**Comment by edsko** [comment]
Closed in #120.
